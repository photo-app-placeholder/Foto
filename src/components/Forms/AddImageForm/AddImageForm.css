.imageform {
  display: flex;
  flex-direction: column;
  width: 180px;
  margin: auto;
}

.imageform select {
  height: 30px;
  width: 200px;
}

.imageform div {
  width: 100%;
  align-items: left;
  margin-bottom: 40px;
  margin-top: 20px;
}

.imageform button {
  width: 100px;
  height: 20px;
}
