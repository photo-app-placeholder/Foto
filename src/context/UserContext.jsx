import { createContext, useContext, useState } from 'react';
import { getUser } from '../services/users';

export const UserContext = createContext();

export const UserProvider = ({ children }) => {
  const currentUser = getUser();
  const [user, setUser] = useState(
    currentUser ? { id: currentUser.id, email: currentUser.email } : {}
  );

  console.log(user);

  const contextValue = { user, setUser };

  return (
    <UserContext.Provider value={contextValue}>{children}</UserContext.Provider>
  );
};

export const useUser = () => {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('Error: useUser needs to be used within UserContext');
  }
  return context;
};
