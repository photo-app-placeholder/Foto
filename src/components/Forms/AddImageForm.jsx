import React from 'react';
import { useState, useEffect } from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { client } from '../../services/client';
import { useUser } from '../../context/UserContext';
import { fetchAlbumsByUser } from '../../services/albums';
import { uploadPhoto } from '../../services/photos';
import profileHook from '../../hooks/profileHook';
import AlbumView from '../../views/Album/AlbumView';

export default function AddImageForm() {
  const [albums, setAlbums] = useState([]);
  const [album, setAlbum] = useState({});
  const [photo, setPhoto] = useState('');
  const [caption, setCaption] = useState('');
  const history = useHistory();
  const { profile } = profileHook();
  console.log(album);
  const { user } = useUser();

  useEffect(() => {
    const fetchData = async () => {
      const data = await fetchAlbumsByUser(user.id);
      setAlbums(data);
    };
    fetchData();
  }, []);

  const handleSubmit = async (e) => {
    try {
      e.preventDefault();
      await uploadPhoto({
        caption: caption,
        photo: photo,
        user_id: user.id,
        album: album.title,
        album_id: album.id,
      });
      alert(`your photo has been uploaded to ${album.title}`);
      history.replace(`/${user.id}/${album.title}`);
    } catch {
      throw new Error('something went wrong uploading your image');
    }
  };

  const handleUpload = async (e) => {
    e.preventDefault();
    try {
      if (!e.target.files || e.target.files === 0) {
        throw new Error('please choose a file to upload');
      }
      const file = e.target.files[0];
      const fileExt = file.name.split('.').pop();
      const fileName = `${new Date().toISOString()}.${fileExt}`;
      const filePath = `${user.id}/${album.title}/${fileName}`;
      setPhoto(
        `https://enluotcdncgmywquucnd.supabase.in/storage/v1/object/public/photos/${filePath}`
      );
      let { error: uploadError } = await client.storage
        .from('photos')
        .upload(filePath, file);
      if (uploadError) {
        throw uploadError;
      }
    } catch (error) {
      alert('error uploading');
    }
  };

  return (
    <form>
      <select onChange={(e) => setAlbum(JSON.parse(e.target.value))}>
        <option>Pick an album</option>
        {albums.map((albumOption) => (
          <option key={albumOption.id} value={JSON.stringify(albumOption)}>
            {albumOption.title}
          </option>
        ))}
      </select>
      <input
        required
        type="file"
        accept="image/*"
        id="single"
        onChange={(e) => {
          handleUpload(e);
        }}
      />
      <input
        type="text"
        value={caption}
        onChange={(e) => setCaption(e.target.value)}
      />
      <button onClick={handleSubmit}>Upload</button>
    </form>
  );
}
