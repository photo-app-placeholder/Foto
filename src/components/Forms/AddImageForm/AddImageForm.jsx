import React from 'react';
import { useState, useEffect } from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { client } from '../../../services/client';
import { useUser } from '../../../context/UserContext';
import { fetchAlbumsByUser } from '../../../services/albums';
import { uploadPhoto } from '../../../services/photos';
import profileHook from '../../../hooks/profileHook';
import styles from './AddImageForm.css';

const { imageform } = styles;

export default function AddImageForm() {
  const [albums, setAlbums] = useState([]);
  const [album, setAlbum] = useState({});
  const [photo, setPhoto] = useState('');
  const [caption, setCaption] = useState('');
  const history = useHistory();
  const { profile } = profileHook();
  const { username } = profile[0];
  const { user } = useUser();

  useEffect(() => {
    const fetchData = async () => {
      const data = await fetchAlbumsByUser(user.id);
      setAlbums(data);
      setAlbum(data[0]);
    };
    fetchData();
  }, []);

  const handleSubmit = async (e) => {
    e.preventDefault();
    const input = document.getElementById('single');
    const file = input.files[0];
    const filePath = photo.split('photos/').pop();
    try {
      await uploadPhoto({
        caption: caption,
        photo: photo,
        user_id: user.id,
        album: album.title,
        album_id: album.id,
        private_public: album.private_public,
        username: username,
      });
      let { error: uploadError } = await client.storage
        .from('photos')
        .upload(filePath, file);
      if (uploadError) {
        throw uploadError;
      }
      alert(`your photo has been uploaded to ${album.title}`);
      if (album.private_public === true) {
        history.replace(`/${username}/${album.id}/unlock`);
      } else {
        history.replace(`/${username}/${album.id}`);
      }
    } catch {
      throw new Error('something went wrong uploading your image');
    }
  };

  const handleUpload = async (e) => {
    e.preventDefault();

    if (!e.target.files || e.target.files === 0) {
      throw new Error('please choose a file to upload');
    }
    const file = e.target.files[0];
    const fileExt = file.name.split('.').pop();
    const fileName = `${new Date().toISOString()}.${fileExt}`;
    const filePath = `${user.id}/${album.title}/${fileName}`;
    setPhoto(
      `https://enluotcdncgmywquucnd.supabase.in/storage/v1/object/public/photos/${filePath}`
    );
  };

  const handleAlbumSelect = (e) => {
    e.preventDefault();
    setAlbum(JSON.parse(e.target.value));
    const input = document.getElementById('single');
    const file = input.files[0];
    const fileExt = file.name.split('.').pop();
    const fileName = `${new Date().toISOString()}.${fileExt}`;
    const filePath = `${user.id}/${album.title}/${fileName}`;
    setPhoto(
      `https://enluotcdncgmywquucnd.supabase.in/storage/v1/object/public/photos/${filePath}`
    );
  };

  return (
    <div className={imageform}>
      <form>
        <div>
          <select onChange={handleAlbumSelect}>
            {albums.map((albumOption) => (
              <option key={albumOption.id} value={JSON.stringify(albumOption)}>
                {albumOption.title}
              </option>
            ))}
          </select>
        </div>
        <>
          <div>
            <input
              required
              type="file"
              accept="image/*"
              id="single"
              onChange={handleUpload}
            />
          </div>
          <div>
            <textarea
              type="text"
              value={caption}
              placeholder="Caption"
              onChange={(e) => setCaption(e.target.value)}
            />
          </div>
          <button onClick={handleSubmit}>Upload</button>
        </>
      </form>
    </div>
  );
}
