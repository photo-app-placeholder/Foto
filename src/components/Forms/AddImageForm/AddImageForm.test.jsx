import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../../context/UserContext';
import AddImageForm from './AddImageForm';

test('should display a form to add images', () => {
  const { container } = render(
    <UserProvider>
      <MemoryRouter>
        <AddImageForm />
      </MemoryRouter>
    </UserProvider>
  );
  expect(container).toMatchSnapshot();
});
