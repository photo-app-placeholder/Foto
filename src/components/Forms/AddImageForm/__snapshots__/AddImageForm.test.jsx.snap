// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display a form to add images 1`] = `
<div>
  <div
    class="imageform"
  >
    <form>
      <div>
        <select />
      </div>
      <div>
        <input
          accept="image/*"
          id="single"
          required=""
          type="file"
        />
      </div>
      <div>
        <textarea
          placeholder="Caption"
          type="text"
        />
      </div>
      <button>
        Upload
      </button>
    </form>
  </div>
</div>
`;
